src/main/java/build/buildfarm/worker/Executor.java

```diff
@@ -0,0 +1,244 @@
+      stdoutSink = worker.instance.getStreamOutput(stdoutStreamName);
+    } else {
+      stdoutSink = nullOutputStream;
+    }
+    if (stderrStreamName != null && !stderrStreamName.isEmpty() && worker.config.getStreamStderr()) {
+      stderrSink = worker.instance.getStreamOutput(stderrStreamName);
+    } else {
+      stderrSink = nullOutputStream;
+    }
+
+    long startNanoTime = System.nanoTime();
+    int exitValue = -1;
+    Process process;
+    try {
+      process = processBuilder.start();
+      process.getOutputStream().close();
+    } catch(IOException ex) {
+      ex.printStackTrace();
+      // again, should we do something else here??
+      ActionResult.Builder resultBuilder = ActionResult.newBuilder()
+          .setExitCode(exitValue);
+      return resultBuilder;
+    }
+
+    InputStream stdoutStream = process.getInputStream();
+    InputStream stderrStream = process.getErrorStream();
+
+    ByteStringSinkReader stdoutReader = new ByteStringSinkReader(
+        process.getInputStream(), stdoutSink);
+    ByteStringSinkReader stderrReader = new ByteStringSinkReader(
+        process.getErrorStream(), stderrSink);
+
+    Thread stdoutReaderThread = new Thread(stdoutReader);
+    Thread stderrReaderThread = new Thread(stderrReader);
+    stdoutReaderThread.start();
+    stderrReaderThread.start();
+
+    boolean doneWaiting = false;
+    if (timeout == null) {
+      exitValue = process.waitFor();
+    } else {
+      while (!doneWaiting) {
+        long timeoutNanos = timeout.getSeconds() * 1000000000L + timeout.getNanos();
+        long remainingNanoTime = timeoutNanos - (System.nanoTime() - startNanoTime);
+        if (remainingNanoTime > 0) {
+          if (process.waitFor(remainingNanoTime, TimeUnit.NANOSECONDS)) {
+            exitValue = process.exitValue();
+            doneWaiting = true;
+          }
+        } else {
+          process.destroyForcibly();
+          process.waitFor(100, TimeUnit.MILLISECONDS); // fair trade, i think
+          doneWaiting = true;
+        }
+      }
+    }
+    if (!stdoutReader.isComplete()) {
+      stdoutReaderThread.interrupt();
+    }
+    stdoutReaderThread.join();
+    if (!stderrReader.isComplete()) {
+      stderrReaderThread.interrupt();
+    }
+    stderrReaderThread.join();
+    ActionResult.Builder resultBuilder = ActionResult.newBuilder()
+        .setExitCode(exitValue);
+    ByteString stdoutRaw = stdoutReader.getData();
+    if (stdoutRaw.size() > 0) {
+      CASInsertionControl control = worker.config.getStdoutCasControl();
+      boolean withinLimit = stdoutRaw.size() <= control.getLimit();
+      if (withinLimit) {
+        resultBuilder.setStdoutRaw(stdoutRaw);
+      }
+      if (control.getPolicy() == CASInsertionControl.Policy.ALWAYS_INSERT ||
+          (!withinLimit && control.getPolicy() == CASInsertionControl.Policy.INSERT_ABOVE_LIMIT)) {
+        resultBuilder.setStdoutDigest(worker.instance.putBlob(stdoutRaw));
```

Comment:
So I guess this is an exception to the stage separation concept -- you upload stuff here from the "wrong" stage? I guess you could separate the stages completely, but then you'd need to save the stdout/err as "special" files, which might not be worth it, or pass them around in some other way.

Reply:
This is definitely an issue - I will have to transport these outputs through the context to the report result stage, good catch.